### PR TITLE
test(tile-select): expanding snapshot coverage

### PR DESCRIPTION
**Pull request by eriklharper**
**Related Issue:** #3570 

## Summary

This PR does a full expansion (135 in total) of snapshot coverage for `calcite-tile-select`, taking into account each prop in isolation and as they relate to other props and also minimizing redundancy across snapshots as much as possible.

I've also paid attention not to break the full width fixes in #1966.  I added full snapshot coverage for full width display both with and without the input.

These are my "snapshot notes" which allow me to document all the snapshots that make sense to capture based on looking at each prop in isolation.  The items bolded correspond to actual snapshots taken.

- Checked
    - Off
        - **Icon + Heading + Description**
    - On
        - **Icon + Heading + Description**
- **Description**
    - **With Heading**
    - **With Icon**
    - **With Heading + Icon**
- Disabled
    - Heading + Description + Icon
        - Input Enabled
            - On
                - Checked
                    - Off
                        - **Radio**
                            - **Hover**
                            - **Focus**
                            - **Click**
                        - **Checkbox**
                            - **Hover**
                            - **Focus**
                            - **Click**
                    - On
                        - **Radio**
                            - **Hover**
                            - **Focus**
                            - **Click**
                        - **Checkbox**
                            - **Hover**
                            - **Focus**
                            - **Click**
            - Off
                - Checked
                    - **Off**
                        - **Hover**
                        - **Focus**
                        - **Click**
                    - **On**
                        - **Hover**
                        - **Focus**
                        - **Click**
- Heading
    - Input Enabled
        - **Off**
        - **On**
- **Hidden**
- Icon
    - Input Enabled
        - **Off**
            - **With Heading (Large Visual)**
        - **On**
            - **With Heading (Large Visual)**
- Input Alignment
    - End
        - Input Enabled
            - On
                - **Icon**
                - **Heading**
                - **Description**
                - **Icon + Heading**
                - **Icon + Description**
                - **Icon + Heading + Description**
                - **Heading + Description**
- Input Enabled
    - On
        - **Icon**
        - **Description**
        - **Icon + Heading**
        - **Icon + Description**
        - **Icon + Heading + Description**
        - **Heading + Description**
    - Off
        - **Icon**
        - **Description**
        - **Icon + Heading**
        - **Icon + Description**
        - **Icon + Heading + Description**
        - **Heading + Description**
- RTL
    - Input Enabled
        - Off
            - **Icon**
            - **Heading**
            - **Description**
            - **Icon + Heading**
            - **Icon + Description**
            - **Icon + Heading + Description**
            - **Heading + Description**
        - On
            - Input Alignment
                - Start
                    - **Icon**
                    - **Heading**
                    - **Description**
                    - **Icon + Heading**
                    - **Icon + Description**
                    - **Icon + Heading + Description**
                    - **Heading + Description**
                - End
                    - **Icon**
                    - **Heading**
                    - **Description**
                    - **Icon + Heading**
                    - **Icon + Description**
                    - **Icon + Heading + Description**
                    - **Heading + Description**
- Theme
    - Light
        - Heading + Description + Icon
            - Input Enabled
                - On
                    - Checked
                        - **On**
                            - **Hover**
                            - **Focus**
                            - **Click**
                        - **Off**
                            - **Hover**
                            - **Focus**
                            - **Click**
                - Off
                    - Checked
                        - **On**
                            - **Hover**
                            - **Focus**
                            - **Click**
                        - **Off**
                            - **Hover**
                            - **Focus**
                            - **Click**
    - Dark
        - Disabled
            - Off
                - Heading + Description + Icon
                    - Input Enabled
                        - On
                            - Checked
                                - **On**
                                    - **Hover**
                                    - **Focus**
                                    - **Click**
                                - **Off**
                                    - **Hover**
                                    - **Focus**
                                    - **Click**
                        - Off
                            - Checked
                                - **On**
                                    - **Hover**
                                    - **Focus**
                                    - **Click**
                                - **Off**
                                    - **Hover**
                                    - **Focus**
                                    - **Click**
            - On
                - Heading + Description + Icon
                    - Input Enabled
                        - On
                            - Checked
                                - **On**
                                    - **Hover**
                                    - **Focus**
                                    - **Click**
                                - **Off**
                                    - **Hover**
                                    - **Focus**
                                    - **Click**
                        - Off
                            - Checked
                                - **On**
                                    - **Hover**
                                    - **Focus**
                                    - **Click**
                                - **Off**
                                    - **Hover**
                                    - **Focus**
                                    - **Click**
- Width
    - Full
        - **Icon**
        - **Heading**
        - **Description**
        - **Icon + Heading**
        - **Icon + Description**
        - **Icon + Heading + Description**
        - **Heading + Description**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
...
...25880299)) ([#3485](https://github.com/Esri/calcite-components/issues/3485)) ([bf33843](https://github.com/Esri/calcite-components/commit/bf33843689cc437e1e5c4c5d3f91164de3c99603)) ([#3484](https://github.com/Esri/calcite-components/issues/3484)) ([8e91ad9](https://github.com/Esri/calcite-components/commit/8e91ad9afa2f7cf8eca95d09e93dbfcddd212a18)) ([#3483](https://github.com/Esri/calcite-components/issues/3483)) ([cc0004a](https://github.com/Esri/calcite-components/commit/cc0004a00f25ebf7b3881f9ebae414b807165105)) ([#3482](https://github.com/Esri/calcite-components/issues/3482)) ([0c5e2bc](https://github.com/Esri/calcite-components/commit/0c5e2bce3da7d55ef1205208b1e3d99263e9ecc4)) ([#3477](https://github.com/Esri/calcite-components/issues/3477)) ([2d61415](https://github.com/Esri/calcite-components/commit/2d6141518584043becab5fa1e523e22dc011dac8)) ([#3471](https://github.com/Esri/calcite-components/issues/3471)) ([a949003](https://github.com/Esri/calcite-components/commit/a949003b06fa8089ff25d28cc16c71479e193e2d)) ([#3498](https://github.com/Esri/calcite-components/issues/3498)) ([109f15a](https://github.com/Esri/calcite-components/commit/109f15aef2282ce2027ada35b3d04f422b4e9f08))
```
Comment (jcfranco) [inline CHANGELOG.md:15]: Can you revert this unrelated change?

File: src/components/calcite-radio-button/calcite-radio-button.tsx [hunk 2]
```diff
@@ -237,6 +239,13 @@ export class CalciteRadioButton implements LabelableComponent, CheckableFormComp
     });
   }
 
+  private getTabIndex(): number {
+    if (this.disabled) {
+      return;
```
Comment (jcfranco) [inline src/components/calcite-radio-button/calcite-radio-button.tsx:244]: I think this should return -1. Can probably be included in the return statement below too.

File: src/components/calcite-tile-select/calcite-tile-select.stories.ts [hunk 3]
```diff
@@ -1,76 +1,1880 @@
...
+          { name: "icon", value: icon },
+          { name: "heading", value: heading },
+          { name: "description", value: description }
+        ]
+      })
+    )
+    .rtl()
+    .snapshot("RTL Input Enabled Off Icon + Heading + Description")
+
+    // RTL Input Enabled Off Heading + Description
```
Comment (jcfranco) [inline src/components/calcite-tile-select/calcite-tile-select.stories.ts:854]: Since `snapshot` has the name, can you remove these comments?

File: src/components/calcite-tile-select/calcite-tile-select.stories.ts [hunk 4]
```diff
@@ -1,76 +1,1880 @@
...
+          { name: "width", value: "full" }
+        ]
+      })
+    )
+    .executeScript(toggleCentered)
+    .snapshot("Input Enabled Icon + Description")
+
+    // Width Full Input Enabled Icon + Heading + Description
+    .executeScript(
+      setKnobs({
```
Comment (jcfranco) [inline src/components/calcite-tile-select/calcite-tile-select.stories.ts:1852]: This is most impressive! Great job adding all of these specific tests. 🧪

File: src/components/calcite-checkbox/calcite-checkbox.tsx [hunk 5]
```diff
@@ -213,7 +213,7 @@ export class CalciteCheckbox implements LabelableComponent, CheckableFormCompome
           onFocus={this.onToggleFocus}
           ref={(toggleEl) => (this.toggleEl = toggleEl)}
           role="checkbox"
-          tabIndex={0}
+          tabIndex={this.disabled ? undefined : 0}
```
Comment (jcfranco) [inline src/components/calcite-checkbox/calcite-checkbox.tsx:216]: `undefined` should be `-1`, otherwise it will get converted to `0`. The prop expects a number.

File: src/components/calcite-radio-button/calcite-radio-button.tsx [hunk 6]
```diff
@@ -352,8 +361,10 @@ export class CalciteRadioButton implements LabelableComponent, CheckableFormComp
   };
 
   private onContainerFocus = (): void => {
-    this.focused = true;
-    this.calciteInternalRadioButtonFocus.emit();
+    if (!this.disabled) {
```
Comment (jcfranco) [inline src/components/calcite-radio-button/calcite-radio-button.tsx:364]: Can you submit a separate PR with the `disabled` behavior-related changes (i.e., no styles)? You already did all of the heavy lifting, so I can take care of adding tests.

File: src/demos/calcite-tile-select.html [hunk 7]
```diff
@@ -645,5 +559,91 @@
         </calcite-tile-select-group>
       </div>
     </div>
+
+    <!-- A horizonal line-->
+    <hr />
```
Comment (jcfranco) [inline src/demos/calcite-tile-select.html:564]: Can this be split into a separate PR? It's technically not related to snapshot coverage. 

I should add that no action should be needed in this PR for the separated-out code/PRs after these are installed and `master` is merged into this one.

**Review by jcfranco — COMMENTED** [review]
This is looking great, @eriklharper!

File: src/components/calcite-radio-button/calcite-radio-button.tsx (hunk 6 above)
Comment (eriklharper) [reply src/components/calcite-radio-button/calcite-radio-button.tsx:364]: If I do that, the accepted new focus snapshots will show up incorrect on the next Screener run.  Unless we're ok with accepting invalid visual states in order to merge this and then accept the new states in a follow-up PR?

File: src/demos/calcite-tile-select.html (hunk 7 above)
Comment (eriklharper) [reply src/demos/calcite-tile-select.html:564]: I'll revert this.  I moved the `calcite-tile` examples down to the bottom since they're not the main focus of the demo, but this isn't important for the goal of this branch.

File: src/components/calcite-checkbox/calcite-checkbox.tsx (hunk 5 above)
Comment (eriklharper) [reply src/components/calcite-checkbox/calcite-checkbox.tsx:216]: Do we want disabled inputs to be focusable via clicks?

The intention here is to not set `tabindex` at all when the input is disabled, which is what I think we want in the case of a disabled inputs.  If not, then clicks on the disabled input will focus it, and thus apply focus styles which don't make sense if it is disabled.

File: src/components/calcite-tile-select/calcite-tile-select.stories.ts (hunk 4 above)
Comment (eriklharper) [reply src/components/calcite-tile-select/calcite-tile-select.stories.ts:1852]: Thank ya sir!  Definitely a lot of work, but well worth it.  We shouldn't miss any more visual regressions with this component any more :)

File: src/components/calcite-radio-button/calcite-radio-button.tsx (hunk 2 above)
Comment (eriklharper) [reply src/components/calcite-radio-button/calcite-radio-button.tsx:244]: See my earlier question about disabled inputs.

File: src/components/calcite-checkbox/calcite-checkbox.tsx (hunk 5 above)
Comment (jcfranco) [reply src/components/calcite-checkbox/calcite-checkbox.tsx:216]: Ignore my comment, you are right. It looks like Stencil is smart enough to remove the attribute when setting the prop to `undefined`. This is different when working with the raw element itself where the only way to do that was to set the attribute (not prop) to "" or remove it manually (see https://codepen.io/jcfranco/pen/mdBBvvv?editors=1010).

File: src/components/calcite-radio-button/calcite-radio-button.tsx (hunk 6 above)
Comment (jcfranco) [reply src/components/calcite-radio-button/calcite-radio-button.tsx:364]: My suggestion was to hold on this one until the separate disabled PR lands (w/o styling changes). That wouldn't impact the visual states here, right?

File: src/components/calcite-radio-button/calcite-radio-button.tsx (hunk 2 above)
Comment (jcfranco) [reply src/components/calcite-radio-button/calcite-radio-button.tsx:244]: Based on the updated thread on the earlier question, can you explicitly return `undefined` and update the return type?

File: src/components/calcite-radio-button/calcite-radio-button.tsx (hunk 6 above)
Comment (eriklharper) [reply src/components/calcite-radio-button/calcite-radio-button.tsx:364]: It will affect the disabled focus visual states.  Basically before this change, a disabled checkbox or radio-button would display a focus halo, and this corrects it so that when focusing a tile-select with `input-enabled` and `disabled` both on it won't show the focus halo in that specific case.

Comment (jcfranco) [reply src/components/calcite-radio-button/calcite-radio-button.tsx:364]: Gotcha. I'm fine ignoring visual diffs from the other PR since this would get installed shortly after.

Comment (eriklharper) [reply src/components/calcite-radio-button/calcite-radio-button.tsx:364]: Here are the fixes for this in a new PR: https://github.com/Esri/calcite-components/pull/3784

Comment (eriklharper) [reply src/components/calcite-radio-button/calcite-radio-button.tsx:364]: Once that's merged I can update this branch that way we don't have to accept any incorrect snapshots.

**Review by jcfranco — APPROVED** [review]
💥
